### ISSUE TITLE
design sketch for Quick App tile

### ISSUE THREAD

**Issue by adriatic** [monterey-framework/monterey#175]
**_Note: this idea should be used to solve the issue [make better use of GistRun in the context of Monterey](https://github.com/monterey-framework/monterey/issues/165). It is kept separate as it is a larger concept than GistRun use in Monterey)_**
***

**Comment by adriatic** [comment]
I really like the quick setup needed to make a small sample app and run it using the GistRun app. **[This GistRun](https://gist.run/?id=809bf843b45efdf8589179f88fd1e507)** used to create the "live problem reports* is a good example of such situation.

Ideally, I would like us to have the "embedded" version of GistRun in Monterey at some point but at the moment, I think that we should get by with a **collection of prebuilt gists**  similar to the **[Aurelia KendoUI component - problem report](https://gist.github.com/JeroenVinke/809bf843b45efdf8589179f88fd1e507)**. 

Each of the gists from the above defined collection is called a **Quick app skeleton** and each of them is specifically built for a given application type. The type is an object that contains the application ***language*** (Javascript / Typescript), ***platform*** (node / asp.net) and ***loader*** (SystemJS, Webpack, RequireJS).

I will address the Quick App UX issues separately - at the moment just assume that at the time the user clicks on the **`Quick App`** tile, Monterey "knows" what type of the application wants to create. The following actions should take place:


1. Start the browser (the type of the browser is defined in user preferences) with the clone of a requested Quick app ***template***

**Comment by JeroenVinke** [comment]
I think it's useful for us to check out the aurelia [quick start guide](http://aurelia.io/hub.html#/doc/article/aurelia/framework/latest/quick-start/1) since that references aurelia via a CDN. Is this something we can use?

Also, while monterey does know what your project look like, that may not be what you want to use in a gist. So we need a screen for this where monterey provides defaults.

**Comment by adriatic** [comment]
### Somehow, I have the feeling that my above post is not understood.
Quick App is meant to be a "RAD" like tool, which should be able to create and run a gist that contains almost any application type that Monterey supports now. It almost indicates the Quick App should be called Quick Monterey should be "parallel to" and not "part of" Monterey"

I would like to prototype this idea (as an Monterey branch) - but as I will be very slow doing this (Monterey is still a secret for me as far as the internals go), I would like to first finish the current testing and get this one out.

**Comment by JeroenVinke** [comment]
> I would like to prototype this idea (as an Monterey branch) - but as I will be very slow doing this (Monterey is still a secret for me as far as the internals go)

Instead of prototyping a running version of this, perhaps it's quicker/easier to create mockups first (https://moqups.com/#!/)

**Comment by adriatic** [comment]
That is what I will do, yes.

**Comment by adriatic** [comment]
Let me start with the a lot more specific approach - beginning with the GistID which is presented to GitRun tool as `https://gist.run/?id=809bf843b45efdf8589179f88fd1e507`.

This gist is really a collection of files (screenshot below)

![image](https://cloud.githubusercontent.com/assets/2712405/17824399/c8aec07c-662f-11e6-84e5-232d592eafa0.png)

where `app.html` and `app.js` are "prebuilt" Aurelia Component that the user is expected to customize, and `index.html` and `main.js` have all needed information that this Aurelia Component needs to be "built and run" - with all the references to remote resources defined in `index.html`, including our own bundle

```
https://rawgit.com/aurelia-ui-toolkits/aurelia-kendoui-bundles/1.0.0-beta.1.0.6/config2.js"
```

this gist is a complete definition of a mini Aurelia App consisting of a single, empty component

```
<!-- put your view here  -->
<template>
  <p>Loaded</p>
</template>
```

and
```
/*  put your view model code here */
export class App {
  
}
```

As we know from the samples in KendoUI catalog app, these small applications could be quite impressive and all you need to do to make them alive is to push on the **`Run`** button:

![image](https://cloud.githubusercontent.com/assets/2712405/17824975/126ae29c-6633-11e6-99ad-fe81be977d48.png)

***

***This describes the specific case: writing and running a mini Aurelia-KendoUI bridge based application in a nice cozy environment - almost like writing Basic apps.***

### What defines the nice, cozy environment, we need to build?

**Comment by adriatic** [comment]
### What are Quick App types?

In a sense - sky is the limit. Starting from using this tool to create problem reports using Monterey to using using this feature as a beginning of a RAD IDE :smile: .

### What defines the nice, cozy environment, we need to build?

In order to ensure that Monterey remains relatively lightweight we should continue hosting all run-time components at places where they live now, despite the obvious fact that Quick App would be a lot more responsive if the bundles were local. This perhaps calls for the preference where the bundles are downloaded once and kept local - in which case we need to ensure the support for keeping them current.

In summary - we need to provide as many different bundles as many Quick Apps we want to support (KendoUI based, Materialize based, etc)

It is also important to make a good judgment about the effort it would take to be able to "teach" Monterey to use any existing Quick App (read a given gist) and expand it further to become a "normal" app. 

### Open issues

- I am not sure that I fully understand what does GistRun do when the user clicks on the **`Run`** button (see [this](https://github.com/gist-run/gist-run/blob/master/src/worker/worker-activator.js))

**Comment by adriatic** [comment]
### Still missing

Quick App UX
...

**Comment by JeroenVinke** [comment]
While I like this idea, one problem has been mentioned very briefly, but needs to be addressed with more detail. Hopefully this will make Quick apps less limited and more powerful.

So we know that the biggest limitation of gist.run is the way packages are loaded. For the kendo bridge we have an external SystemJS bundle that we can reference from index.html. This bundle contains everything we might want to use in a gist.

But if we wanted to use MomentJS in this gist (which is not in the bundle), then we would have a problem. We would need to update the external bundle so that it includes MomentJS. But there is another, much easier way. We can edit the `index.html` and use a script tag to reference MomentJS from a CDN: `<script src="https://cdnjs.cloudflare.com/ajax/libs/moment.js/2.14.1/moment.min.js"></script>`. With this script tag in `index.html` we can use MomentJS in the gist.

Now comes the thought behind my comment:
> I think it's useful for us to check out the aurelia quick start guide since that references aurelia via a CDN. Is this something we can use?

I only know that this quick start guide uses aurelia via a CDN. I do not know how this works, but if we can load aurelia via a script tag (`<script src="http://some.cdn/aurelia.js"></script>`) then Aurelia can be loaded in the same way as MomentJS. 

So if we can do that, then we can create a UI where the user can say "I want Aurelia, Kendo, and I want MomentJS". When Monterey creates the quick app (the app.html, app.js and index.html files), the index.html will be:

```
<html>
<head>
<script src="http://some.cdn/aurelia.js"></script>
<script src="http://some.cdn/kendo.all.min.js"></script>
<script src="http://some.cdn/moment.js"></script>
</head>

<body>
</body>
</html>
```

Take a look at https://cdnjs.com/. With CDNJS you can search for a library and copy the CDN URL into your app. Now, in the UI where users select the libraries to use for a quick app, we can use cdnjs as a datasource. You would be able to search for, and choose "MomentJS" but also any other library that is on cdnjs.

We're not done yet, as many aurelia plugins will not be hosting their plugins on a CDN (aurelia-kendoui-bridge included). So this is something we need to think about some more.

**Comment by adriatic** [comment]
>> I think it's useful for us to check out the aurelia quick start guide since that references aurelia via a CDN. Is this something we can use?

[Aurelia Quick Start](http://aurelia.io/hub.html#/doc/article/aurelia/framework/latest/quick-start/1) does not expect any CDN-like hosted aurelia component. Instead the reader is asked to fetch this [zip file](http://aurelia.io/downloads/basic-aurelia-project.zip) which contains:

![image](https://cloud.githubusercontent.com/assets/2712405/17832013/20892e9e-66c7-11e6-9ab6-e16e9b69f08a.png)

so all reeferences to aurelia framework are "project-local".

I did not find any CDN (including `cdnjs`, where there is one single month-old request for aurelia).

There are several OSS (free) CDN hosts.

I will discuss the Quick App tile idea with Ron (unless you disagree with that)

**Comment by adriatic** [comment]
@JeroenVinke writes [above](https://github.com/monterey-framework/monterey/issues/175#issuecomment-241202245)

> So we know that the biggest limitation of gist.run is the way packages are loaded. For the kendo bridge gist we have an external SystemJS bundle that we can reference from index.html. This bundle contains everything we use in our gists.

> But if we wanted to use MomentJS in this gist (which is not in the bundle), then we would have a problem. We would need to update the external bundle so that it includes MomentJS. But there is another, much easier way. We can edit the `index.html` and use a script tag to reference MomentJS from a CDN: `<script src="https://cdnjs.cloudflare.com/ajax/libs/moment.js/2.14.1/moment.min.js"></script>`. With this script tag in `index.html` we can use MomentJS in the gist.

In order to provide the continuity of this discussion, I propose to do it here, regardless of the long thread. Once we have a conclusion, I will extract the relevant part and make a new condensed version of this document.
